[4004] Integrate item selector into dialogs

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4004

aachavarria committed:
changes on DialogHeader subtitle to support JSX.Element;
Changes on emptyState to support children;
adding emptyState to CompareVersionsDialog.tsx;

aachavarria committed:
Prettier format;

rart commented:
Please integrate with New content dialog, Dependencies dialog, History & View Versions too.

aachavarria committed:
DialogHeader changes;
PathNavigator: isSelectMode optional;
adding resetVersionsState to Widget;
adding ItemsStateProps to Redux Store, Epics, Actions, Reducer...;
Adding PathNavigator to SingleItemSelector.tsx;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

 Conflicts:
	ui/app/src/components/Dialogs/DialogHeader.tsx

rart commented on ui/app/src/state/reducers/items.ts:
```diff
@@ -0,0 +1,52 @@
+  isFetching: null,
+  error: null,
+  items: null,
+  path: 'site/website'
```
Path should be at widget state. This should just be an item container.

rart commented on ui/app/src/state/reducers/items.ts:
```diff
@@ -0,0 +1,52 @@
+    [fetchPathComplete.type]: (state, { payload }) => ({
+      ...state,
+      byId: {...state.byId, ...createLookupTable(payload)},
+      items: payload.map(item => item.id),
```
Belongs to widget.

rart commented on ui/app/src/models/Item.ts:
```diff
@@ -89,3 +92,11 @@ export interface LegacyItem {
+  error: ApiResponse;
+  isFetching: boolean;
+  items: string[],
+  path: string
```
Items and path belongs to widgets

aachavarria committed:
adding consumers;
adding Breadcrumbs to SingleItemSelector.tsx;
adding addconsumer and fetchChildrenByPath to useEffect on CompareVersionsDialog.tsx;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

 Conflicts:
	ui/app/src/modules/Content/History/CompareVersionsDialog.tsx

aachavarria committed:
quickFixes on Merge;

aachavarria committed:
PathNavigatorBreadcrumbs.tsx: optional onMenu;
PathNavigatorItem.tsx: optional onOpenItemMenu;
PathNavigatorList.tsx: optionl, onSelectItem and onOpenItemMenu;
Widget.tsx: moving itemsFromPath to utils;
GlobalDialogManager.tsx: adding itemsBranch;
adding BreadCrumb to SingleItemSelector.tsx;

aachavarria committed:
adding selected item to Consumer interface;
moving Breadcrumbs outsite of suspense;
validations on useEffect;

aachavarria committed:
adding leafs;
WIP

rart commented on ui/app/src/models/Item.ts:
```diff
@@ -89,3 +93,21 @@ export interface LegacyItem {
   [prop: string]: any;
 }
+
+export interface Consumer extends PaginationOptions {
```
Move this out

rart commented on ui/app/src/modules/Content/Authoring/SingleItemSelector.tsx:
```diff
@@ -14,142 +14,166 @@
-        <Typography variant={labelVariant || 'body1'}>{selectItem.label}</Typography>
-      </div>
+      <Typography
+        variant={titleVariant || 'body1'}
```
set default on the prop

rart commented on ui/app/src/modules/Content/Authoring/SingleItemSelector.tsx:
```diff
@@ -14,142 +14,166 @@
+        selectedItem &&
+        <div className={classes.selectedItem}>
+          <ItemIcon className={clsx(classes.itemIcon, propClasses?.itemIcon)} />
+          <Typography variant={labelVariant || 'body1'}>{selectedItem.label}</Typography>
```
default on the prop

rart commented on ui/app/src/modules/Content/Authoring/SingleItemSelector.tsx:
```diff
@@ -14,142 +14,166 @@
+      >
+        <Breadcrumbs
+          keyword={consumer?.keywords}
+          breadcrumb={consumer ? consumer?.breadcrumb : []}
```
consumer?.breadcrumb ?? []

rart commented on ui/app/src/modules/Content/Authoring/SingleItemSelector.tsx:
```diff
@@ -14,142 +14,166 @@
+        />
+        <SuspenseWithEmptyState resource={itemsResource}>
+          <PathNavigatorList
+            leafs={consumer ? consumer?.leafs : []}
```
consumer?.leafs ?? []

aachavarria committed:
changes on addConsumer();

aachavarria committed:
PR Feedback

aachavarria committed:
integrating SingleItemSelector on HistoryDialog.tsx

aachavarria committed:
removing noUsed Classes

aachavarria committed:
integrating ItemSelector to NewContentDialog.tsx

aachavarria committed:
Deleting itemBranch, Moving reducer to SingleItemSelector.tsx;
Deleting Consumer.ts;
Deleting itemStateProps;
Removing Single selector from all dialogs;
Making parse to parseLegacyItemToSandBoxItem;

rart commented on ui/app/src/components/ComponentMenu.tsx:
```diff
@@ -117,9 +117,8 @@ export default function ComponentMenu(props: ComponentMenuProps) {
-        }));
+        const path = models[modelId].craftercms.path || models[parentId].craftercms.path;
+        dispatch(fetchItemVersions({ path }));
         dispatch(showHistoryDialog());
```
Use batchActions where ever required

rart commented on ui/app/src/modules/Content/History/CompareVersionsDialog.tsx:
```diff
@@ -95,35 +113,38 @@ export default function CompareVersionsDialog(props: CompareVersionsDialogProps)
+          compareVersionsBranch.compareVersions &&
+          contentTypesBranch.byId &&
+          !compareVersionsBranch.isFetching &&
+          !contentTypesBranch.isFetching
```
With all the &&, the boolean constructor doesn’t make sense

aachavarria committed:
adding item and rootPath to CompareVersionsDialog.tsx;
adding forkJon to SingleItemSelector.tsx to get parentsPath;

aachavarria committed:
SingleItemSelector.tsx: changes to use parentPath on fetchParentsItemsComplete()

aachavarria committed:
remove console.log()

aachavarria committed:
adding selectedItem and RootPath to SingleItemSelector.tsx;
adding keepMounted to bothDialogs to avoid issue with Resource being cached;
Adding getItem service;
Adding actions to changeDialogCurrentItem;

aachavarria committed:
adding interface for CompareVersionsBranch;
adding useMemo on Resource creation;
removing oldIDs;

aachavarria committed:
changes on actions to use a Object;
Changes on NewContentDialog to receive a Item;
Addind ItemSelector on NewContentDialog.tsx;
adding a default to rootPath;

aachavarria committed:
removing console.log

aachavarria committed:
adding Item Selector To DependenciesDialog.tsx;

aachavarria committed:
ComponentMenu Changes;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

 Conflicts:
	ui/app/src/components/ComponentMenu.tsx
	ui/app/src/services/content.ts
	ui/app/src/state/actions/dialogs.ts
	ui/app/src/state/reducers/dialogs/dependencies.ts
	ui/app/src/state/reducers/dialogs/newContent.ts

aachavarria committed:
Merge Changes

aachavarria committed:
pagination fix

rart commented on ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx:
```diff
@@ -343,12 +346,38 @@ function DependenciesDialogUI(props: DependenciesDialogUIProps) {
+          {/*    setOpenSelector(false);*/}
+          {/*    dispatch(fetchItemVersions({ path: item.path }));*/}
+          {/*  }}*/}
+          {/*/>*/}
```
x

rart commented on ui/app/src/services/content.ts:
```diff
@@ -77,6 +77,12 @@ export function getLegacyItem(site: string, path: string): Observable<LegacyItem
   );
 }
 
+export function getItem(site: string, path: string): Observable<SandboxItem> {
```
Rename to getSandboxItem

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -184,22 +184,23 @@ export function useStateResource<ReturnType = unknown, SourceType = unknown, Err
+  // Purposely not adding checkers on to the effect dependencies to avoid
+  // consumer re-renders to trigger the effect. `checkers` should be taken
+  // as a "initialValue" sort of param. The functions should not mutate
+  // throughout the component lifecycle.
```
The effect is now in sync with it's dependency. Remove comment.

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -68,11 +93,27 @@ export function withIndex(path: string): string {
+  return splitPath.map( function(value, i) {
+    return '/' + splitPath.slice(1, i + 1).join('/');
+  }).splice(2);
+}
```
Format

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -68,11 +93,27 @@ export function withIndex(path: string): string {
   withoutIndex,
-  withIndex
+  withIndex,
+  getParentsFromPath
```
Add all new methods

aachavarria committed:
adding SingleItemSelector to DependenciesDialog.tsx

aachavarria committed:
adding SingleItemSelector to DependenciesDialog.tsx

aachavarria committed:
Rename get getItem to getSandboxItem

aachavarria committed:
removing comment

aachavarria committed:
format and adding new methods to export default

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

aachavarria committed:
adding historyDialog to sideBarPanel;
adding dependenciesDialog to Widget.tsx;
General fixes on HistoryDialog and CompareVersionsDialog.tsx;

aachavarria committed:
sending a sandBox Item to dependenciesDialog.tsx;

aachavarria committed:
admin-configuration.js: changes on the item uri;
common-api.js: adding showHistoryDialog to viewContentHistory;
deleting history.js;
adding revertPath to versions( used when historyDialogConfig;

aachavarria committed:
changes on layered, wcm to add rootPath;
changes on isNavigable;
changes on getParentsFromPath();

aachavarria committed:
fixes in SingleItemSelector.tsx to support path === rootPath

aachavarria committed:
Creating DialogBase;
fetchItemVersions should receive rootpath and item;
Changes onGlobalDialogManager to remove Rootpath and Item from HistoryDialogs;
Adding onClosed to HistoryDialogs, reducer, actions;

aachavarria committed:
adding resetVersionsState when HistoryDialogClosed

aachavarria committed:
Changes on resetVersionsState epic

aachavarria committed:
adding (closed and DialogBase) to confirm and Error Dialog;

aachavarria committed:
Doing changes on DeleteDialog.tsx, PublishDialog.tsx, DependenciesDialog.tsx to support SandBoxItems;
Changes on common-api.js to send sandboxItems;
Adding onClosed to DeleteDialog.tsx, PublishDialog.tsx, DependenciesDialog.tsx, NewContentDialog.tsx;

aachavarria committed:
fix

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

aachavarria committed:
adding useDialogOnClosed(props.onClosed) to all dialogs;
removing DialogBase.tsx;

aachavarria committed:
adding batchActions()

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -721,7 +721,7 @@ var nodeOpen = false,
           payload: {
             open: true,
-            items,
+            items: items.map(item => CrafterCMSNext.services.content.parseLegacyItemToSandBoxItem(item)),
```
Move to utils/content

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -748,84 +748,61 @@ var nodeOpen = false,
-          },
-          true
-        );
+        CrafterCMSNext.system.store.dispatch({ type: 'SHOW_HISTORY_DIALOG' });
```
use batchActions

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -748,84 +748,61 @@ var nodeOpen = false,
+              payload: { id: eventIdOnClose }
+            }
+          }
+        });
```
batchActions

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -748,84 +748,61 @@ var nodeOpen = false,
+          document.dispatchEvent(eventNS);
+          amplify.publish('HISTORY_REVERT');
+          // end of TODO
+          CrafterCMSNext.system.store.dispatch({ type: 'RESET_VERSIONS_STATE' });
```
This was no longer necessary, was it?

rart commented on ui/app/src/components/ComponentMenu.tsx:
```diff
@@ -99,43 +87,41 @@ export default function ComponentMenu(props: ComponentMenuProps) {
-          path: models[modelId].craftercms.path || models[parentId].craftercms.path
+        dispatch(fetchItemVersions({ item: item }));
+        dispatch(showHistoryDialog({
+          //onRevert
```
x

rart replied:
batchactions

rart commented on ui/app/src/components/Dialogs/ConfirmDialog.tsx:
```diff
@@ -24,6 +23,8 @@ import Button from '@material-ui/core/Button';
 import { defineMessages, useIntl } from 'react-intl';
 import StandardAction from '../../models/StandardAction';
+import { Dialog } from '@material-ui/core';
+import { useDialogOnClosed } from '../../utils/hooks';
```
Let’s rename to useOnUnmount

rart commented on ui/app/src/components/SystemStatus/GlobalDialogManager.tsx:
```diff
@@ -152,7 +161,9 @@ function GlobalDialogManager() {
       <HistoryDialog
         open={state.history.open}
         versionsBranch={versionsBranch}
+        //onRevert={createCallback(state.history.onRevert, dispatch)}
```
x

rart commented on ui/app/src/models/Version.ts:
```diff
@@ -14,7 +14,7 @@
  */
 
-import { LegacyItem } from './Item';
+import { LegacyItem, SandboxItem } from './Item';
```
Check terminal warnings. Are you still using LegacyItem here for example?

aachavarria replied:
yes

rart commented on ui/app/src/modules/Content/Authoring/NewContentDialog.tsx:
```diff
@@ -44,6 +43,11 @@ import { SuspenseWithEmptyState } from '../../../components/SystemStatus/Suspenc
+import SingleItemSelector from './SingleItemSelector';
+import { SandboxItem } from '../../../models/Item';
+import { showErrorDialog } from '../../../state/reducers/dialogs/error';
+import { Dialog } from '@material-ui/core';
```
Make sure to import the dialog from its file and not from core (check all over the dialogs where you did the find & replace)

rart replied:
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/Authoring/NewContentDialog.tsx:
```diff
@@ -187,20 +188,36 @@ function ContentTypesGrid(props: ContentTypesGridProps) {
+    <Dialog
+      open={props.open}
+      onClose={props.onClose}
+
```
x empty line

rart commented on ui/app/src/modules/Content/Delete/DeleteDialog.tsx:
```diff
@@ -64,19 +65,21 @@ interface DeleteDialogUIProps {
-    onSuccess?(response?: any): any;
-  }
+  onClose(): any;
+  onClosed(): any;
```
needn’t be mandatory

rart replied:
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/History/CompareVersionsDialog.tsx:
```diff
@@ -70,136 +84,194 @@ interface CompareVersionsDialogProps extends CompareVersionsDialogBaseProps {
+    <Dialog
+      open={props.open}
+      onClose={props.onClose}
+
```
x

rart commented on ui/app/src/modules/Content/History/CompareVersionsDialog.tsx:
```diff
@@ -70,136 +84,194 @@ interface CompareVersionsDialogProps extends CompareVersionsDialogBaseProps {
+      open={props.open}
+      onClose={props.onClose}
+
+      fullWidth={true}
```
use shortcut syntax

rart commented on ui/app/src/modules/Content/History/ViewVersionDialog.tsx:
```diff
@@ -122,25 +126,41 @@ interface ViewVersionDialogProps extends ViewVersionDialogBaseProps {
+      open={props.open}
+      onClose={props.onClose}
 
-  const resource = useStateResource<Resource, ViewVersionDialogProps>(props, {
```
Is there a pointless empty line here? remove if so.

rart commented on ui/app/src/modules/Content/History/HistoryDialog.tsx:
```diff
@@ -240,70 +264,60 @@ export default function HistoryDialog(props: HistoryDialogProps) {
       })
     );
+    dispatch(fetchContentTypes());
+    dispatch(fetchContentVersion({ path, versionNumber: version.versionNumber }));
```
batch actions

rart replied:
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/Publish/PublishDialog.tsx:
```diff
@@ -405,21 +405,32 @@ function PublishDialogUI(props: PublishDialogUIProps) {
+    <Dialog
+      open={props.open}
+      onClose={props.onClose}
 
```
x

rart commented on ui/app/src/state/epics/dialogs.ts:
```diff
@@ -85,7 +85,7 @@ export default [
         const onClose = getDialogState(type, state)?.onClose;
         return [
-          createClosedAction(type),
+          //createClosedAction(type),
```
x

rart commented on ui/app/src/state/reducers/dialogs/confirm.ts:
```diff
@@ -22,17 +22,13 @@ export default createReducer<GlobalState['dialogs']['confirm']>(
     [showConfirmDialog.type]: (state, { payload }) => ({
-      // By default, if no callback is specified, assume an ok button.
-      // To not have a "Ok" button, action creator must be called with onOk: null.
-      // This allows easily sending information dialogs with a ok button.
```
why remove comment? this is still the case. put back?

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -248,3 +245,10 @@ export function useMinimizeDialog(initialTab: MinimizedDialog) {
 
   return state?.minimized ?? initialTab.minimized;
 }
+export const useDialogOnClosed = (onClosed) => {
```
rename to useOnUnmount

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -248,3 +245,10 @@ export function useMinimizeDialog(initialTab: MinimizedDialog) {
   return state?.minimized ?? initialTab.minimized;
 }
+export const useDialogOnClosed = (onClosed) => {
+  useOnMount(() => {
```
Use useeffect directly to avoid the unnecessary abstraction

rart reviewed (changes requested):
Please address comments

aachavarria committed:
moving parseLegacyItemToSandBoxItem to utils;

aachavarria committed:
using BATCH_ACTIONS on Common-api.js

aachavarria committed:
adding BATCH_ACTIONS to common-api.js

aachavarria committed:
adding more BATCH_ACTIONS to common-api.js

aachavarria committed:
PRFeedback;

aachavarria committed:
revertComment;

aachavarria committed:
removing empty space;

aachavarria committed:
renaming;

aachavarria committed:
renaming;

rart committed:
Sugar syntax

rart commented on ui/app/src/components/Dialogs/ConfirmDialog.tsx:
```diff
@@ -24,6 +23,8 @@ import Button from '@material-ui/core/Button';
 import React, { PropsWithChildren } from 'react';
 import { defineMessages, useIntl } from 'react-intl';
 import StandardAction from '../../models/StandardAction';
+import { Dialog } from '@material-ui/core';
```
Still incorrect. Import from their individual files.

rart commented on ui/app/src/components/SystemStatus/ErrorDialog.tsx:
```diff
@@ -14,14 +14,14 @@
 import IconButton from '@material-ui/core/IconButton';
 import CloseIcon from '@material-ui/icons/CloseRounded';
-import { createStyles, makeStyles, Theme } from '@material-ui/core';
+import { createStyles, Dialog, makeStyles, Theme } from '@material-ui/core';
```
Incorrect. Import from their individual files.

rart commented on ui/app/src/modules/Content/Delete/DeleteDialog.tsx:
```diff
@@ -243,17 +252,17 @@ export default function DeleteDialog(props: DeleteDialogProps) {
 
   useEffect(() => {
     open &&
```
We no longer need this, right? 👆🏻

rart commented on ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx:
```diff
@@ -534,41 +502,57 @@ function DependenciesDialogUI(props: DependenciesDialogUIProps) {
+    <Dialog
+      open={props.open}
+      onClose={props.onClose}
+      fullWidth={true}
```
Use shortcut syntax

rart commented on ui/app/src/modules/Content/History/CompareVersionsDialog.tsx:
```diff
@@ -15,40 +15,43 @@
 import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMoreRounded';
+import SingleItemSelector from '../Authoring/SingleItemSelector';
+import { Dialog } from '@material-ui/core';
```
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx:
```diff
@@ -55,6 +52,8 @@ import Checkbox from '@material-ui/core/Checkbox';
 import EmbeddedLegacyEditors from '../../Preview/EmbeddedLegacyEditors';
 import { ApiResponse } from '../../../models/ApiResponse';
+import SingleItemSelector from '../Authoring/SingleItemSelector';
+import { Dialog } from '@material-ui/core';
```
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/Delete/DeleteDialog.tsx:
```diff
@@ -32,26 +33,26 @@ import createStyles from '@material-ui/core/styles/createStyles';
 import DialogFooter from '../../../components/Dialogs/DialogFooter';
 import Button from '@material-ui/core/Button';
 import CircularProgress from '@material-ui/core/CircularProgress';
+import { Dialog } from '@material-ui/core';
```
Still pending 👆🏻

rart commented on ui/app/src/modules/Content/History/HistoryDialog.tsx:
```diff
@@ -47,6 +46,9 @@ import {
   showViewVersionDialog
 } from '../../../state/actions/dialogs';
+import SingleItemSelector from '../Authoring/SingleItemSelector';
+import { Dialog } from '@material-ui/core';
```
Still pending 👆🏻

rart committed:
Cleanup

rart committed:
Cleanup

rart committed:
Cleanup

rart reviewed (changes requested):
Missed a few comments from last review. **Make sure** you click on "Load hidden" — from _this_ and _previous_ review — to resolve all.
<img width="646" alt="image" src="https://user-images.githubusercontent.com/3928341/81386253-fee4c380-9114-11ea-9840-e8a40abe2265.png">

rart commented:
Please solve conflicts

aachavarria committed:
Feedback

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4004

 Conflicts:
	ui/app/src/components/ComponentMenu.tsx
	ui/app/src/modules/Content/Authoring/NewContentDialog.tsx
	ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx
	ui/app/src/modules/Preview/QuickCreate.tsx

aachavarria committed:
onClose, onClosed, onDismiss changes();

aachavarria committed:
merge issues

aachavarria committed:
merge issues

aachavarria committed:
wrong action creator

rart committed:
Remove disable backdrop click
